### ISSUE TITLE
Rule proposal: Require to return something when we detect a Promise

### ISSUE THREAD

**Issue by jfmengels** [avajs/eslint-plugin-ava#154]
I'm seeing this kind of error more and more in our codebase at work
```js
test('foo', t => {
  foo.bar()
  .then(baz => {
  });
});
```

The problem being that we use Promises in this test and that we don't return it, meaning that the test will in most cases pass even if the assertions in the `.then` should fail later in the execution.

Therefore I'm proposing a rule (`no-unreturned-promise`?) that reports an error when a test contains a Promise, but does not return anything. This should not affect callback tests `test.cb(...)` or tests with an async function.

We already have code to [detect Promises](https://github.com/avajs/eslint-plugin-ava/blob/master/rules/prefer-async-await.js#L5) (should we also add support for detecting `catch`?), so we could use that (and move it to [`eslint-ast-utils`](https://github.com/jfmengels/eslint-ast-utils) later on ;) ).

### Invalid

```js
test('foo', t => {
  foo.bar()
  .then(baz => {
    t.fail(); // should fail but won't
  });
});
```

### Valid

```js
test('foo', t => {
  return foo.bar() // was returned
  .then(baz => {
  });
});

test('foo', t => {
  foo.bar(); // no promises
});

test.cb('foo', t => { // callback test
  foo.bar()
  .then(baz => {
  });
});

test('foo', async t => { // async implementation
  foo.bar()
  .then(baz => {
  });
});
```

**Comment by sindresorhus** [comment]
I considered this some months ago, but I just use `await` for everything now, so I didn't see the need. Why are you not just using `await` instead of returning the promise?

> should we also add support for detecting catch?

👍

**Comment by jfmengels** [comment]
Actually, in this case, I saw this in a Mocha test.

Not everybody in my team is up to date with using async functions, and some still use Promises in their tests. But even if they were, we could have gotten the error by forgetting to return the Promise when converting a function to test (and therfore also the test function) from synchronous to asynchronous using Promises or from callbacks to Promises.

Yesterday, I found AVA tests using Promises that were not returned and were actually silently failing (though, in those cases, there were no `.then`, so this would not be detected here anyway). Just this morning, I found Mocha tests (we still use Mocha in some older projects) that had this detectable behavior. I intend to open an issue in eslint-plugin-mocha very soonish for a similar rule.

I think it's a nice safety net to have anyway, as it's a nasty trap to fall in and often hard to detect.
Note: As soon as we'll be able to know the types of expressions with Flow/TS in ESLint, this would be an awesome rule and we'd get a lot less errors. Fingers crossed :D

**Comment by sindresorhus** [comment]
👍 Oh, yeah, totally agree about this rule. Was just curious.

**Comment by novemberborn** [comment]
As an aside, Node v6 now reports unhandled rejections, we could fail a test file if they occur?

**Comment by jfmengels** [comment]
That sounds good, but would that work if the test is already considered as finished and passing?

**Comment by novemberborn** [comment]
The test run will fail, even if all the tests pass.

I believe in a future Node release unhandled rejections will behave like `error` events without listeners and crash the process.
